File: BookWorm-NET/BookWormNET/Services/Implementation/TransactionPdfService.cs
using BookWormNET.Models;
using iText.IO.Font.Constants;
using iText.Kernel.Font;
using iText.Kernel.Pdf;
using iText.Layout.Element;
using iText.Layout.Properties;
using iText.Layout;


namespace BookWormNET.Services.Implementation
{
    public class TransactionPdfService
    {
        public byte[] GenerateInvoice(Transaction transaction,
                                      List<TransactionItem> items)
        {
            using var stream = new MemoryStream();

            PdfWriter writer = new PdfWriter(stream);
            PdfDocument pdf = new PdfDocument(writer);
            Document document = new Document(pdf);

            PdfFont titleFont = PdfFontFactory.CreateFont(StandardFonts.HELVETICA_BOLD);
            PdfFont normalFont = PdfFontFactory.CreateFont(StandardFonts.HELVETICA);

            // Title
            document.Add(new Paragraph("Bookworm - Transaction Invoice")
                .SetFont(titleFont)
                .SetFontSize(18)
                .SetTextAlignment(TextAlignment.CENTER));

            document.Add(new Paragraph(" "));

            // Transaction Details
            document.Add(new Paragraph($"Transaction ID: {transaction.TransactionId}")
                .SetFont(normalFont));

            document.Add(new Paragraph($"User: {transaction.User.UserName}")
                .SetFont(normalFont));

            document.Add(new Paragraph($"Type: {transaction.TransactionType}")
                .SetFont(normalFont));

            document.Add(new Paragraph($"Status: {transaction.Status}")
                .SetFont(normalFont));

            document.Add(new Paragraph($"Date: {transaction.CreatedAt}")
                .SetFont(normalFont));

            document.Add(new Paragraph(" "));

            // Table (4 columns)
            Table table = new Table(4).UseAllAvailableWidth();

            table.AddHeaderCell("Book");
            table.AddHeaderCell("Price");
            table.AddHeaderCell("Qty");
            table.AddHeaderCell("Total");

            foreach (var item in items)
            {
                decimal lineTotal = (item!.Price ?? 0m) * (item.Quantity ?? 0);

                table.AddCell(item.Product?.ProductName ?? "N/A");
                table.AddCell("₹" + (item.Price ?? 0m));
                table.AddCell(item.Quantity.ToString());
                table.AddCell("₹" + lineTotal);
            }


            document.Add(table);

            document.Add(new Paragraph(" "));
            document.Add(new Paragraph("Total Amount: ₹" + transaction.TotalAmount)
                .SetFont(titleFont));

            document.Close();

            return stream.ToArray();
        }
    }
}

File: BOOKWORM_PRO/Bookworm/src/main/java/com/example/Services/TransactionPdfService.java
package com.example.Services;

import com.example.models.Transaction;
import com.example.models.TransactionItem;
import com.lowagie.text.*;
import com.lowagie.text.pdf.*;

import org.springframework.stereotype.Service;

import java.io.ByteArrayOutputStream;
import java.math.BigDecimal;
import java.util.List;

@Service
public class TransactionPdfService {

    public byte[] generateInvoice(Transaction transaction,
                                  List<TransactionItem> items) {

        ByteArrayOutputStream out = new ByteArrayOutputStream();

        Document document = new Document(PageSize.A4);
        PdfWriter.getInstance(document, out);

        document.open();

        Font titleFont = FontFactory.getFont(FontFactory.HELVETICA_BOLD, 18);
        Font normalFont = FontFactory.getFont(FontFactory.HELVETICA, 12);

        // Title
        document.add(new Paragraph("Bookworm - Transaction Invoice", titleFont));
        document.add(new Paragraph(" "));

        // Transaction details
        document.add(new Paragraph("Transaction ID: " + transaction.getTransactionId(), normalFont));
        document.add(new Paragraph("User: " + transaction.getUser().getUserName(), normalFont));
        document.add(new Paragraph("Type: " + transaction.getTransactionType(), normalFont));
        document.add(new Paragraph("Status: " + transaction.getStatus(), normalFont));
        document.add(new Paragraph("Date: " + transaction.getCreatedAt(), normalFont));
        document.add(new Paragraph(" "));

        // Table
        PdfPTable table = new PdfPTable(4);
        table.setWidthPercentage(100);

        table.addCell("Book");
        table.addCell("Price");
        table.addCell("Qty");
        table.addCell("Total");

        for (TransactionItem item : items) {
            table.addCell(item.getProduct().getProductName());
            table.addCell("₹" + item.getPrice());
            table.addCell(String.valueOf(item.getQuantity()));
            //table.addCell("₹" + (item.getPrice() * item.getQuantity()));
            BigDecimal lineTotal =
                    item.getPrice().multiply(BigDecimal.valueOf(item.getQuantity()));

            table.addCell("₹" + lineTotal);

        }

        document.add(table);

        document.add(new Paragraph(" "));
        document.add(new Paragraph("Total Amount: ₹" + transaction.getTotalAmount(), titleFont));

        document.close();

        return out.toByteArray();
    }
}